src/Bedrock.Framework.Experimental/Protocols/Kafka/KafkaMessageWriter.cs

```diff
@@ -0,0 +1,55 @@
+﻿#nullable enable
+
+using Bedrock.Framework.Experimental.Protocols.Kafka.Messages.Requests;
+using Bedrock.Framework.Experimental.Protocols.Kafka.Services;
+using Bedrock.Framework.Infrastructure;
+using Bedrock.Framework.Protocols;
+using Microsoft.Extensions.Logging;
+using System.Buffers;
+
+namespace Bedrock.Framework.Experimental.Protocols.Kafka
+{
+    public class KafkaMessageWriter : IMessageWriter<KafkaRequest>
+    {
+        private readonly IMessageCorrelator correlator;
+        private readonly ILogger<KafkaMessageWriter> logger;
+
+        public KafkaMessageWriter(
+            IMessageCorrelator messageCorrelator,
+            ILogger<KafkaMessageWriter> logger)
+        {
+            this.correlator = messageCorrelator;
+            this.logger = logger;
+        }
+
+        public void WriteMessage(KafkaRequest message, IBufferWriter<byte> output)
+        {
+            var correlationId = this.correlator.GetCorrelationId(message);
+            var writer = new BufferWriter<IBufferWriter<byte>>(output);
+            var clientId = message.ClientId;
+
+            // On every outgoing message...
+            var pw = new PayloadWriter(isBigEndian: true)
+                .StartCalculatingSize("payloadSize")
```

Comment:
First byte in outgoing messages is the size.
PayloadWriter will track byteswritten and store off a Memory<byte> at this location. Once ```EndSizeCalculation``` is called it will write the size to the stored Memory<byte>.